File: bms/go/ach/moov_io_ach_validator_CalculateCheckDigit/moov_io_ach_validator_CalculateCheckDigit.go
package test

import (
	"math"
	"unicode/utf8"
)

var ()

const (
	moov_io_ach_GLPrenoteDebit = 48
)

func (v *moov_io_ach_validator) CalculateCheckDigit(routingNumber string) int {
	return moov_io_ach_CalculateCheckDigit(routingNumber)
}

func moov_io_ach_CalculateCheckDigit(routingNumber string) int {
	if n := utf8.RuneCountInString(routingNumber); n != 8 && n != 9 {
		return -1
	}

	var sum int
	for i, r := range routingNumber {
		// Don't process check digit of routing number
		if i >= 8 {
			break
		}

		// Reject anything that's not a digit
		if r < '0' || r > '9' {
			return -1 // only digits are allowed
		}

		// Calculate the check digit
		var n int32 = (r - '0')

		switch i {
		case 0, 3, 6:
			sum += int(n * 3)
		case 1, 4, 7:
			sum += int(n * 7)
		case 2, 5:
			sum += int(n)
		}
	}

	return moov_io_ach_roundUp10(sum) - sum
}

func moov_io_ach_roundUp10(n int) int {
	return int(math.Ceil(float64(n)/10.0)) * 10
}

type moov_io_ach_validator struct{}

File: translations_only/success/go/ach/moov_io_ach_validator_CalculateCheckDigit/claude2/CAPR/moov_io_ach_validator_CalculateCheckDigit.rs

use std::char;

const MOOV_IO_ACH_GLPRENOTEDEBIT: i32 = 48;

struct MoovIoAchValidator;

impl MoovIoAchValidator {
    pub fn calculate_check_digit(&self, routing_number: &str) -> i32 {
        moov_io_ach_calculate_check_digit(routing_number)
    }
}

fn moov_io_ach_calculate_check_digit(routing_number: &str) -> i32 {
    let n = routing_number.len();
    if n != 8 && n != 9 {
        return -1;
    }

    let mut sum = 0;
    for (i, c) in routing_number.chars().enumerate() {
        if i >= 8 {
            break;
        }

        if c < '0' || c > '9' {
            return -1;
        }

        let n = (c as i32) - ('0' as i32);

        match i {
            0 | 3 | 6 => sum += n * 3,
            1 | 4 | 7 => sum += n * 7,
            2 | 5 => sum += n,
            _ => (),
        }
    }

    let rounded = moov_io_ach_round_up_10(sum);
    rounded - sum
}

fn moov_io_ach_round_up_10(n: i32) -> i32 {
    ((n as f64 / 10.0).ceil() * 10.0) as i32
}

